streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java

```diff
@@ -58,7 +112,88 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareActiveTask
         final Map<ProcessId, KafkaStreamsAssignment> kafkaStreamsAssignments,
         final SortedSet<TaskId> tasks
     ) {
-        throw new UnsupportedOperationException("Not Implemented.");
+        if (tasks.isEmpty()) {
+            return kafkaStreamsAssignments;
+        }
+
+        if (!hasValidRackInformation(applicationState)) {
```

Comment:
It seems a little silly to have to run this exact check twice, which we are essentially doing by throwing an exception if it fails. For example the StickyTaskAssignor runs this check and if false, skips calling the rack-aware optimization. 

Imo, to best keep the current behavior where it simply skips applying this optimization if it doesn't have all of the rack id info it needs, we should just return early here instead of throwing an IllegalStateException, and not call it at all in the StickyTaskAssignor (or any of the assignors). The point is to move any rack-related logic out of the assignors and encapsulate it in these utility methods, so they can simply be called without knowing anything about how rack aware assignment works. Throwing an exception if you call this method without first having run some pre-check that users won't easily know about seems a bit cruel